### PR TITLE
2521 Fetch listed token and save to a storage

### PR DESCRIPTION
**Pull request by halink0803**

### CODE REVIEW

File: accounting/cmd/accounting-listed-token-fetcher/main.go [hunk 1]
```diff
@@ -10,14 +10,16 @@ import (
 	"github.com/urfave/cli"
 
 	listedtoken "github.com/KyberNetwork/reserve-stats/accounting/listed-token-fetcher"
+	listedtokenstorage "github.com/KyberNetwork/reserve-stats/accounting/listed_token_storage"
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: refactored.

File: accounting/cmd/accounting-listed-token-fetcher/main.go [hunk 2]
```diff
@@ -10,14 +10,16 @@ import (
...
 	"github.com/KyberNetwork/reserve-stats/lib/blockchain"
 	"github.com/KyberNetwork/reserve-stats/lib/etherscan"
 )
 
 const (
-	blockFlag          = "block"
-	reserveAddressFlag = "reserve-address"
+	blockFlag           = "block"
+	reserveAddressFlag  = "reserve-address"
+	defaultAccountingDB = "accounting"
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: changed

File: accounting/cmd/accounting-listed-token-fetcher/main.go [hunk 3]
```diff
@@ -86,7 +89,16 @@ func run(c *cli.Context) error {
...
+	db, err := libapp.NewDBFromContext(c)
+	if err != nil {
+		return err
+	}
+	listedTokenStorage, err := listedtokenstorage.NewDB(sugar, db)
+	if err != nil {
+		return err
+	}
+
+	fetcher := listedtoken.NewListedTokenFetcher(ethClient, resolv, sugar, listedTokenStorage)
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: changed

File: accounting/listed-token-fetcher/fetcher.go [hunk 4]
```diff
@@ -113,5 +116,6 @@ func (f *Fetcher) GetListedToken(block *big.Int, reserveAddr ethereum.Address,
 
 	// currently print out to cli, save to storage later
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed-token-fetcher/fetcher.go]: Removed

File: accounting/listed_token_storage/listed_token_storage_postgres.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+	tx, err := db.Beginx()
+	if err != nil {
+		return nil, err
+	}
+
+	defer pgsql.CommitOrRollback(tx, logger, &err)
+
+	logger.Debug("initializing database schema")
+	if _, err = tx.Exec(fmt.Sprintf(schemaFmt, tokenTable, tokenTable)); err != nil {
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed_token_storage/listed_token_storage_postgres.go]: changed

[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed_token_storage/listed_token_storage_postgres.go]: changed

File: accounting/listed_token_storage/listed_token_storage_postgres.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+			token.Timestamp); err != nil {
+			return err
+		}
+
+		if len(token.Old) != 0 {
+			for _, oldToken := range token.Old {
+				if _, err = tx.Exec(upsertOldTokenQuery,
+					oldToken.Address,
+					token.Name,
+					token.Symbol,
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed_token_storage/listed_token_storage_postgres.go]: changed

File: accounting/listed_token_storage/listed_token_storage_postgres.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+	var logger = sugar.With("func", "accounting/storage.NewDB")
+
+	tx, err := db.Beginx()
+	if err != nil {
+		return nil, err
+	}
+
+	defer pgsql.CommitOrRollback(tx, logger, &err)
+
+	logger.Debug("initializing database schema")
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed_token_storage/listed_token_storage_postgres.go]: done

File: accounting/listed_token_storage/listed_token_storage_postgres.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	getQuery := fmt.Sprintf(`SELECT address, name, symbol, cast (extract(epoch from timestamp)*1000 as bigint) as timestamp FROM %[1]s`, tokenTable)
+	logger.Debugw("get tokens query", "query", getQuery)
+
+	if err := ltd.db.Select(&result, getQuery); err != nil {
+		logger.Errorw("error query token", "error", err)
+	}
+
+	logger.Debugw("query result from tokens table", "result", result)
+	return result, nil
+}
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed_token_storage/listed_token_storage_postgres.go]: added close method.

File: accounting/listed_token_storage/listed_token_storage_postgres_test.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+		return nil, err
+	}
+	storage, err := NewDB(sugar, db)
+	if err != nil {
+		return nil, err
+	}
+	return storage, nil
+}
+
+func TestListedTokenStorage(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed_token_storage/listed_token_storage_postgres_test.go]: added teardown.

File: accounting/listed_token_storage/listed_token_storage_postgres.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+	VALUES (
+		$1, 
+		$2, 
+		$3,
+		to_timestamp($4::double precision / 1000),
+		(SELECT id FROM "%[1]s" WHERE symbol = $3)
+	)
+	ON CONFLICT (address) DO NOTHING`,
+		tokenTable)
+
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed_token_storage/listed_token_storage_postgres.go]: changed

File: accounting/listed-token-storage/listed_token_storage_postgres.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+		if _, err = tx.Exec(upsertQuery,
+			token.Address,
+			token.Name,
+			token.Symbol,
+			token.Timestamp,
+			token.Address); err != nil {
+			return err
+		}
+
+		if len(token.Old) != 0 {
```
Comment (cyberliem) [inline accounting/listed-token-storage/listed_token_storage_postgres.go]: Redundant condition checking. For loop will not be execute if token.Old has no member

File: accounting/cmd/accounting-listed-token-fetcher/main.go [hunk 12]
```diff
@@ -86,7 +89,21 @@ func run(c *cli.Context) error {
 
 	resolv := blockchain.NewEtherscanContractTimestampResolver(sugar, etherscanClient)
 
+	db, err := libapp.NewDBFromContext(c)
+	if err != nil {
```
Comment (cyberliem) [inline accounting/cmd/accounting-listed-token-fetcher/main.go]: should defer db.close() here.

File: accounting/cmd/accounting-listed-token-fetcher/main.go [hunk 13]
```diff
@@ -98,6 +98,10 @@ func run(c *cli.Context) error {
 		return err
 	}
 
+	defer func(err *error) {
+		*err = listedTokenStorage.Close()
+	}(&err)
```
Comment (cyberliem) [inline accounting/cmd/accounting-listed-token-fetcher/main.go]: the error is modified here, but does not return I think.

Comment (halink0803) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: Yes, it reference to the error outside, you can see it input is &err. It works like function CommitOrRollback.

File: accounting/cmd/accounting-listed-token-fetcher/main.go (hunk 12 above)
Comment (halink0803) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: added.

File: accounting/listed-token-storage/listed_token_storage_postgres.go (hunk 11 above)
Comment (halink0803) [reply accounting/listed-token-storage/listed_token_storage_postgres.go]: removed.

File: accounting/cmd/accounting-listed-token-fetcher/main.go (hunk 13 above)
Comment (cyberliem) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: if error occur, the err will be modified to catch this error. But this function does not return err, it return https://github.com/KyberNetwork/reserve-stats/pull/227/files#diff-1fb2d13b373a897fbdc446f38f696d41R112.

So to guarantee that the function will return any error if occur, better change those lines to:
```go
     if err= listedTokenStorage.CreateOrUpdate(listedTokens); err!=nil {
        return err
     }
     
     return err
```

Comment (halink0803) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: changed.

File: accounting/listed_token_storage/listed_token_storage_postgres_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cyberliem) [reply accounting/listed_token_storage/listed_token_storage_postgres_test.go]: @favadi is it necessary to implement Option Withcustomtablename to support test?

[the remark this answers is not in the document]
Comment (favadi) [reply accounting/listed_token_storage/listed_token_storage_postgres_test.go]: > is it necessary to implement Option Withcustomtablename to support test?

It is good to have, please point @halink0803 to an example.

File: accounting/cmd/accounting-listed-token-fetcher/main.go (hunk 13 above)
Comment (favadi) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: @halink0803 @cyberliem below is proof that this technique of deferred error handling does not work.

```go
package main

import (
	"errors"
	"log"
)

func handleErr1() error {
	var err error

	defer func(err *error) {
		*err = errors.New("I got an error")
	}(&err)

	return err
}

func handleErr2() (err error) {
	defer func(err *error) {
		*err = errors.New("I got an error")
	}(&err)

	return err
}

func main() {
	log.Printf("error1: err=%v", handleErr1())
	log.Printf("error2: err=%v", handleErr2())
}

// 2019/03/21 10:56:48 error1: err=<nil>
// 2019/03/21 10:56:48 error2: err=I got an error
```

File: accounting/cmd/accounting-listed-token-fetcher/main.go [hunk 14]
```diff
@@ -86,7 +90,35 @@ func run(c *cli.Context) error {
...
+
 	fetcher := listedtoken.NewListedTokenFetcher(ethClient, resolv, sugar)
 
-	return fetcher.GetListedToken(block, reserveAddr, tokenSymbol)
+	listedTokens, err := fetcher.GetListedToken(block, reserveAddr, tokenSymbol)
+	if err != nil {
+		return err
+	}
+
+	if err = listedTokenStorage.CreateOrUpdate(listedTokens); err != nil {
```
Comment (favadi) [inline accounting/cmd/accounting-listed-token-fetcher/main.go]: just `return listedTokenStorage.CreateOrUpdate(listedTokens)`

File: accounting/listed-token-storage/listed_token_storage_postgres.go [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+	ON CONFLICT (address) DO NOTHING`,
+		ltd.tableName)
+
+	logger.Debugw("upsert token", "value", upsertQuery)
+
+	tx, err := ltd.db.Beginx()
+	if err != nil {
+		return err
+	}
+	defer pgsql.CommitOrRollback(tx, logger, &err)
```
Comment (favadi) [inline accounting/listed-token-storage/listed_token_storage_postgres.go]: again, this won't work without named error.

File: accounting/listed-token-storage/listed_token_storage_postgres.go [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+	var (
+		logger = ltd.sugar.With(
+			"func",
+			"accounting/listed-token-storage/listedtokenstorage.GetTokens",
+		)
+		result       []common.ListedToken
+		listedTokens = make(map[string]common.ListedToken)
+	)
+
+	getQuery := fmt.Sprintf(`SELECT address, name, symbol, cast (extract(epoch from timestamp)*1000 as bigint) as timestamp FROM %[1]s`, ltd.tableName)
```
Comment (favadi) [inline accounting/listed-token-storage/listed_token_storage_postgres.go]: should be easier to assemble if we use SELF LEFT JOIN line.

```sql
SELECT address, name, symbol,olds.address,olds.timestamp cast FROM tokens LEFT JOIN tokens as olds ON tokens.id = olds.parent_id.
```

Comment (favadi) [reply accounting/listed-token-storage/listed_token_storage_postgres.go]: anw, now I notice we have to do timestamp <-> epoch conversion, should we made `Timestamp` field of common.ListedToken time.Time?

File: accounting/listed-token-storage/listed_token_storage_postgres.go [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+
+	return &ListedTokenDB{
+		sugar:     sugar,
+		db:        db,
+		tableName: tableName,
+	}, nil
+}
+
+//CreateOrUpdate add or edit an record in the tokens table
+func (ltd *ListedTokenDB) CreateOrUpdate(tokens map[string]common.ListedToken) error {
```
Comment (favadi) [inline accounting/listed-token-storage/listed_token_storage_postgres.go]: please add a test case to prove that this works with old/new token logic. Scenario:

1. input: `common.ListedToken`: {address: 0x1, name: A, symbol: 123}`
2. input: `common.ListedToken`: {address: 0x2, name: A, symbol: 123, oldTokens:[ 0x1]}`

expected behaviour: 0x2 records is now linked with newly created 0x1 record.

File: accounting/listed-token-storage/listed_token_storage_postgres.go (hunk 16 above)
Comment (halink0803) [reply accounting/listed-token-storage/listed_token_storage_postgres.go]: Finally we still need to convert from time to timestamp when response to end user. If it does not have performance problem, I think we can still keep it this way.

Comment (favadi) [reply accounting/listed-token-storage/listed_token_storage_postgres.go]: we should always keep timestamp in time.Time, just for consistent sake.

File: accounting/cmd/accounting-listed-token-fetcher/main.go (hunk 14 above)
Comment (halink0803) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: changed

File: accounting/listed-token-fetcher/fetcher.go [hunk 18]
```diff
@@ -73,45 +74,39 @@ func (f *Fetcher) GetListedToken(block *big.Int, reserveAddr ethereum.Address,
 	logger.Infow("reserve address", "reserve", reserveAddr)
 	reserveContractClient, err := contracts.NewReserve(reserveAddr, f.ethClient)
 	if err != nil {
-		return err
+		return result, err
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed-token-fetcher/fetcher.go]: return nil

File: accounting/listed-token-fetcher/fetcher.go [hunk 19]
```diff
@@ -73,45 +74,39 @@ func (f *Fetcher) GetListedToken(block *big.Int, reserveAddr ethereum.Address,
...
 	}
 	callOpts := &bind.CallOpts{BlockNumber: block}
 	conversionRatesContract, err := reserveContractClient.ConversionRatesContract(callOpts)
 	if err != nil {
-		return err
+		return result, err
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed-token-fetcher/fetcher.go]: return nil

File: accounting/listed-token-fetcher/fetcher.go [hunk 20]
```diff
@@ -73,45 +74,39 @@ func (f *Fetcher) GetListedToken(block *big.Int, reserveAddr ethereum.Address,
...
 	}
 
 	// step 2: get listedTokens from conversionRatesContract
 	conversionRateContractClient, err := contracts.NewConversionRates(conversionRatesContract, f.ethClient)
 	if err != nil {
-		return err
+		return result, err
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed-token-fetcher/fetcher.go]: return nil

File: accounting/listed-token-fetcher/fetcher.go [hunk 21]
```diff
@@ -73,45 +74,39 @@ func (f *Fetcher) GetListedToken(block *big.Int, reserveAddr ethereum.Address,
...
 	}
 	listedTokens, err := conversionRateContractClient.GetListedTokens(callOpts)
 	if err != nil {
-		return err
+		return result, err
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed-token-fetcher/fetcher.go]: return nil

File: accounting/listed-token-fetcher/fetcher.go [hunk 22]
```diff
@@ -73,45 +74,39 @@ func (f *Fetcher) GetListedToken(block *big.Int, reserveAddr ethereum.Address,
...
 	}
 
 	for _, address := range listedTokens {
 		symbol, err := tokenSymbol.Symbol(address)
 		if err != nil {
-			return err
+			return result, err
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed-token-fetcher/fetcher.go]: return nil

File: accounting/listed-token-fetcher/fetcher.go [hunk 23]
```diff
@@ -73,45 +74,39 @@ func (f *Fetcher) GetListedToken(block *big.Int, reserveAddr ethereum.Address,
...
 		}
 		name, err := tokenSymbol.Name(address)
 		if err != nil {
-			return err
+			return result, err
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed-token-fetcher/fetcher.go]: return nil

File: accounting/listed-token-fetcher/fetcher.go [hunk 24]
```diff
@@ -73,45 +74,39 @@ func (f *Fetcher) GetListedToken(block *big.Int, reserveAddr ethereum.Address,
...
 		}
 		timestamp, err := f.contractTimestampResolver.Resolve(address)
 		if err != nil {
-			return err
+			return result, err
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed-token-fetcher/fetcher.go]: return nil

File: accounting/listed-token-storage/interface.go [hunk 25]
```diff
@@ -0,0 +1,9 @@
+package listedtokenstorage
```
[the remark this answers is not in the document]
Comment (halink0803) [reply accounting/listed-token-storage/interface.go]: changed

File: accounting/listed-token-storage/listed_token_storage_postgres.go (hunk 15 above)
Comment (halink0803) [reply accounting/listed-token-storage/listed_token_storage_postgres.go]: using named return error

File: accounting/listed-token-storage/listed_token_storage_postgres.go (hunk 16 above)
Comment (halink0803) [reply accounting/listed-token-storage/listed_token_storage_postgres.go]: Changed to using time.Time

File: accounting/listed-token-storage/listed_token_storage_postgres.go (hunk 17 above)
Comment (halink0803) [reply accounting/listed-token-storage/listed_token_storage_postgres.go]: added test with recommended scenario.

File: accounting/cmd/accounting-listed-token-fetcher/main.go (hunk 13 above)
Comment (favadi) [reply accounting/cmd/accounting-listed-token-fetcher/main.go]: I did the closing twice in e42f085cd072e0f106b360965266952c6003cf30.

**Pull request merged**